test/unit/support.d

```diff
@@ -0,0 +1,20 @@
+module support;
+
+/// UDA used to indicate a function should be run before each test.
+enum beforeEach;
+
+/// UDA used to indicate a function should be run after each test.
+enum afterEach;
+
+/// Returns: the default import paths, i.e. for Phobos and druntime.
+string[] defaultImportPaths()
+{
+    import std.path : buildNormalizedPath, buildPath, dirName;
+
+    enum dlangDir = __FILE_FULL_PATH__.dirName.buildNormalizedPath("..", "..", "..");
```

Comment:
This assumption isn't valid for the LDC and GDC test suite. That's why I recommend using `findImportPaths`.

Reply:
`findImportPaths` didn't work [1], that's why I changed it. And how do I know if it finds the host compiler or the newly built compiler?

[1] https://circleci.com/gh/dlang/dmd/25836

Reply:
The existing test runners assume that path is valid:

https://github.com/dlang/dmd/blob/3fa19783c47c968a832349ce8ca3bb7f43199919/test/Makefile#L97-L98

https://github.com/dlang/dmd/blob/3fa19783c47c968a832349ce8ca3bb7f43199919/test/run.d#L283-L284

Or do I need to use the environment variables?

Reply:
> And how do I know if it finds the host compiler or the newly built compiler?

Hmm, `dmd.conf` gets generated as part of building DMD and you know which compiler it finds as it prefers `DMD`, but it looks like the environment variable `DMD` simply wasn't set.
Alternatively, you could run `parseImportPathsFromConfig` directly like:

```
dmdExecDir = env["DMD"].dirName;
parseImportPathsFromConfig(dmd.buildPath("dmd.conf"), dmdExecDir);
```

Though would need a special case for `sc.ini` (Windows) and still is DMD-only...

> Or do I need to use these environment variables?

Yeah, I would definitely use these variables instead of hard-coding it directly, but even then LDC/GDC might use sth. different.
Maybe you can move these bits into the shared code, s.t. not too much duplication happens?

Reply:
Is the conclusion to use the `DRUNTIME_PATH` and `PHOBOS_PATH` environment variables and not `DMD`

> but even then LDC/GDC might use sth. different.

But they are using the current test runners? So it does work for them?

Reply:
> Hmm, `dmd.conf` gets generated as part of building DMD and you know which compiler it finds as it prefers `DMD`

Currently, at least the make runner, runs all tests with: `-conf=`. What does that actually do?

Reply:
> Currently, at least the make runner, runs all tests with: `-conf=`. What does that actually do?

Makes sure dmd doesn't use any other `dmd.conf` and the manually passed import + link paths to druntime and phobos are the only ones that DMD has on its argument.
In other words: it makes DMD "naked"/"blank".

> But they are using the current test runners? So it does work for them?

LDC seems to use the `Makefile`: https://github.com/ldc-developers/ldc/blob/5c24f60cf942c7fea3e36c04213e45f2af1416cf/tests/d2/CMakeLists.txt#L41, but they overwrite `DFLAGS` with their own auto-generated config file (but at least you have a properly set `DMD` from which you could detect the paths).

GDC seems to use have written its own alternative for the `Makefile` / `run.d` for probably a better integration in their build system:
https://github.com/D-Programming-GDC/gdc/blob/master/gdc/gcc/testsuite/gdc.test/gdc-test.exp

In the GDC case, they seem to build the flags to Druntime/Phobos manually. See e.g.:

https://github.com/D-Programming-GDC/gdc/blob/0cfcf9d68afc58b7a6c4d0aaff6f82b742ecac98/gdc/gcc/testsuite/lib/gdc.exp#L88

Reply:
Fixed, now using `DRUNTIME_PATH` and `PHOBOS_PATH` if they exist.